File: pkg/types/manifest.go
package types

type Manifest struct {
	Title       string        `json:"title"`
	Platforms   []Platfom     `json:"platforms,omitempty"`
	Description string        `json:"description,omitempty"`
	Require     []Requirement `json:"requirements,omitempty"`
	Root        []RootItem    `json:"root,omitempty"`
	Commands    []CommandSpec `json:"commands"`
}

type Platfom string

const (
	PlatformLinux Platfom = "linux"
	PlatformMac   Platfom = "macos"
)

type Requirement struct {
	Name string `json:"name"`
	Link string `json:"link,omitempty"`
}

type RootItem struct {
	Title   string         `json:"title"`
	Command string         `json:"command"`
	Params  map[string]any `json:"params"`
}

type CommandSpec struct {
	Name   string      `json:"name"`
	Title  string      `json:"title"`
	Hidden bool        `json:"hidden,omitempty"`
	Inputs []Input     `json:"params,omitempty"`
	Mode   CommandMode `json:"mode,omitempty"`
}

type CommandMode string

const (
	CommandModeList   CommandMode = "list"
	CommandModeDetail CommandMode = "detail"
	CommandModeTTY    CommandMode = "tty"
	CommandModeSilent CommandMode = "silent"
)

type Input struct {
	Type        InputType `json:"type"`
	Name        string    `json:"name"`
	Title       string    `json:"title"`
	Required    bool      `json:"required,omitempty"`
	Default     any       `json:"default,omitempty"`
	Placeholder string    `json:"placeholder,omitempty"`
	Label       string    `json:"label,omitempty"`
}

type InputType string

const (
	InputText     InputType = "text"
	InputTextArea InputType = "textarea"
	InputPassword InputType = "password"
	InputCheckbox InputType = "checkbox"
	InputNumber   InputType = "number"
)

File: typescript/pkg/manifest.ts
export type Manifest = {
  title: string;
  env?: string[];
  description?: string;
  commands: readonly Command[];
  root?: string;
};

export type Command = {
  name: string;
  hidden?: boolean;
  description?: string;
  params?: readonly Input[];
  mode: "filter" | "search" | "detail" | "silent";
};

export type Input = {
  name: string;
  title: string;
  type: "string" | "number" | "boolean";
  optional?: boolean;
};

type InputMap = {
  string: string;
  number: number;
  boolean: boolean;
};

type CommandName<M extends Manifest> = M["commands"][number]["name"];

type CommandByName<M extends Manifest, N extends CommandName<M>> = Extract<
  M["commands"][number],
  { name: N }
>;

type ParamName<M extends Manifest, N extends CommandName<M>> = NonNullable<
  CommandByName<M, N>["params"]
>[number]["name"];

type ParamByName<
  M extends Manifest,
  N extends CommandName<M>,
  K extends ParamName<M, N>,
> = Extract<NonNullable<CommandByName<M, N>["params"]>[number], { name: K }>;

export type Payload<M extends Manifest> = {
  [N in CommandName<M>]:
  & {
    command: N;
    params: CommandByName<M, N>["params"] extends undefined
    ? Record<string, never>
    : {
      [K in ParamName<M, N>]: ParamByName<M, N, K>["optional"] extends true
      ? InputMap[ParamByName<M, N, K>["type"]] | undefined
      : InputMap[ParamByName<M, N, K>["type"]];
    };
  }
  & (CommandByName<M, N>["mode"] extends "search" ? { query: string }
    : Record<string, never>);
}[CommandName<M>];
